lib/matplotlib/axis.py

```diff
@@ -1568,6 +1568,9 @@ def set_major_formatter(self, formatter):
 
         ACCEPTS: A :class:`~matplotlib.ticker.Formatter` instance
         """
+        if not isinstance(formatter, mticker.Formatter):
```

Comment:
To support duck typing, Python code tries to avoid `isinstance(obj, cls)` in favor of `hasattr(obj, name)`. Not sure about the canonical attributes of methods of locators and formatters, however, I would suggest checks like `if not hasattr(formatter, 'format_data')` and `if not hasattr(locator, 'tick_values')`

Reply:
Does that become an AttributeError in that case?

Reply:
eh, and probably should have been TypeError if not and AttributeError.  Certainly ValueError is incorrect?

Reply:
I'd go for TypeError. By intent, the `hasattr` here is a kind of a loose type checking.